### PR TITLE
User Pages

### PR DESCRIPTION
**Pull request by jagtalon**
There's a lot going on here, but here are the important files:

root/static/css/userpages/UserPage.scss
root/static/js/userpages/UserPage.js
root/static/pages/ddh_up_index.html
root/static/partials/c_instantanswer.html
root/static/partials/c_issue.html

### CODE REVIEW

**Comment by jagtalon** [comment]
We have the necessary markup. All we need to do is to make it use the right data

screen shot 2016-03-30 at 11 04 52 am

**Comment by jagtalon** [comment]
screen shot 2016-03-30 at 2 49 45 pm

**Comment by jagtalon** [comment]
Taking shape:

screen shot 2016-03-31 at 11 51 54 am

File: root/static/pages/ddh_up_index.html [hunk 1]
```diff
@@ -1,2 +1,125 @@
...
+<div class="container">
+
+  <!-- Sidebar -->
+  <aside>
+    <div class="profile card">
+      <img src="{{response.gh_data.avatar_url}}" />
+      <div class="content">
+	<h3>{{user.name}}</h3>
+	<div class="gray">
+	  <a href="http://github.com/{{response.gh_data.login}}">@{{response.gh_data.login}}</a>
```
Comment (GuiltyDolphin) [inline root/static/pages/ddh_up_index.html]: @jagtalon Can we switch this to `https`?

File: views/userpage/index.tx [hunk 2]
```diff
@@ -1,103 +1,125 @@
...
+<div class="container">
+
+  <!-- Sidebar -->
+  <aside>
+    <div class="profile card">
+      <img src="{{response.gh_data.avatar_url}}" />
+      <div class="content">
+	<h3>{{user.name}}</h3>
+	<div class="gray">
+	  <a href="http://github.com/{{response.gh_data.login}}">@{{response.gh_data.login}}</a>
```
Comment (GuiltyDolphin) [inline views/userpage/index.tx]: @jagtalon This one too (to `https`). Is there a reason for having it as `http` in the first place?

File: views/userpage/index.tx [hunk 3]
```diff
@@ -1,103 +1,125 @@
...
+	  <br />{{response.gh_data.location}}
+	  </div>
+
+	</div>
+</div>
+    <div class="topic typeC" ng-repeat="(key, value) in response.topics | orderBy:'-value'">
```
Comment (GuiltyDolphin) [inline views/userpage/index.tx]: It'd be awesome if the topics were clickable to alter which IAs were displaying ('filter by topic').

File: root/static/pages/ddh_up_index.html (hunk 1 above)
Comment (jagtalon) [inline root/static/pages/ddh_up_index.html]: @GuiltyDolphin yup, can do!

**Comment by jagtalon** [comment]
@MariagraziaAlastra @bsstoner I changed the definition of `random` because we want it to return the same value every time the function is called. AngularJS thinks that the model is changing thus triggering an update, and sees that the model has changed again which triggers an update until we reach an infinite loop.

See `infdig` error: http://is.gd/Zbiz6o

File: root/static/pages/ddh_up_index.html [hunk 4]
```diff
@@ -1,2 +1,297 @@
...
+<div class="container">
+
+  <!-- Sidebar -->
+  <aside>
+    <div class="profile card">
+      <img src="{{response.gh_data.avatar_url}}" />
+      <div class="content">
+	<h3>{{response.gh_data.name}}</h3>
+	<div class="gray">
+	  <a href="http://github.com/{{response.gh_data.login}}">@{{response.gh_data.login}}</a>
```
Comment (GuiltyDolphin) [inline root/static/pages/ddh_up_index.html]: @jagtalon Looks like this one is still `http`?

**Comment by MariagraziaAlastra** [comment]
@jagtalon Shouldn't that function trigger only once, on page load?

File: root/static/pages/ddh_up_index.html (hunk 4 above)
Comment (jagtalon) [inline root/static/pages/ddh_up_index.html]: Fixed!

**Comment by jagtalon** [comment]
@MariagraziaAlastra It's inside the controller and from what I understand gets re-rendered when it's not [in sync](https://www.youtube.com/watch?v=Eo-KmOd3i7s) with the model

**Comment by MariagraziaAlastra** [comment]
@jagtalon ok, thanks :smile:

**Comment by MariagraziaAlastra** [comment]
@jagtalon Maybe [one time binding](https://docs.angularjs.org/guide/expression#one-time-binding) would solve this too? I need to make a new PR anyway, so I can try that if you wish

**Comment by jagtalon** [comment]
@MariagraziaAlastra one-time binding seems like it would work too. Anything that would prevent the digest from re-calculation I think is a good solution

**Comment by alohaas** [comment]
@jagtalon here I go!

**Comment by jagtalon** [comment]
@alohaas down the rabbit hole! 🐇

File: root/static/partials/c_instantanswer.html [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<!-- Outside is wrapped within <div class="ia..."> -->
+
+<div class="grid-6">
+  <strong><a href="https://duck.co/ia/view/{{ia.id}}" ng-click="addImg('iapage', ia.id)">{{ia.name}}</a></strong>
```
Comment (alohaas) [inline root/static/partials/c_instantanswer.html]: @jagtalon I just tried to click through to Ben's Markdown Reference IA and it redirected me to the IA index. Maybe the page doesn't exist? maybe the url is being built incorrectly?
screen shot 2016-04-12 at 3 53 31 pm

http://jag.duckduckgo.com:5002/u/guiltydolphin ( Markdownreference is the last IA in his maintaining list)

Comment (jagtalon) [inline root/static/partials/c_instantanswer.html]: @alohaas it looks like the data that I have is out of data--that Markdownreference one has been ghosted and shouldn't show up at all (it's not showing up here: https://duck.co/u/guiltydolphin/). Thanks!

**Comment by alohaas** [comment]
@jagtalon I'm curious to know why you've put the github iconfont files in the `root/static/css/userpages/` directory. Should they not live within a `root/static/fonts/userpages/`directory?

**Comment by jagtalon** [comment]
@alohaas oh no reason. Probably should be in the fonts directory. I could move it to fonts/

**Comment by alohaas** [comment]
@jagtalon I think that would be a better home for them. I would add a `userpages` subdirecty in the `fonts` folder and put them in there to maintain consistency across directories. The more orderly and intentional we can be here, the more it will help our flexibility goals.

**Comment by jagtalon** [comment]
@alohaas cool now it's in the fonts/ dir

File: root/static/pages/ddh_up_index.html [hunk 6]
```diff
@@ -1,2 +1,296 @@
...
+	<div class="gray">
+	  <a href="https://github.com/{{response.gh_data.login}}">@{{response.gh_data.login}}</a>
+	  <br />{{response.gh_data.location}}
+	  </div>
+
+	</div>
+</div>
+    <div class="topic typeC" ng-repeat="topic in topics | orderBy:'-count'">
+      <span class="amt"> {{topic.count}}</span>
+      <span> {{topic.topic}}</span>
```
Comment (alohaas) [inline root/static/pages/ddh_up_index.html]: Mr. Chris has a million topics, but they grow dangerously close to the footer of his page.Add some padding?

screen shot 2016-04-13 at 7 54 35 am

Comment (jagtalon) [inline root/static/pages/ddh_up_index.html]: @alohaas good call. I got it

**Comment by jagtalon** [comment]
@alohaas fixed!

**Comment by alohaas** [comment]
@jagtalon I'm seeing a double scroll bar on firefox. When I removed `overflow-x: hidden` from the body element `body` the second scroll bar disappears. Is there a reason that style was applied? Removing it seems not to break anything.

screen shot 2016-04-13 at 10 32 37 am

**Comment by alohaas** [comment]
@jagtalon the user avatars adjacent to IAs aren't sitting in the center of their circles.

screen shot 2016-04-13 at 10 37 46 am

**Comment by jagtalon** [comment]
@alohaas cool should be fixed!

File: root/static/pages/ddh_up_index.html (hunk 6 above)
Comment (alohaas) [inline root/static/pages/ddh_up_index.html]: @jagtalon Chris only has 21 topics in his list, which is more than most contributors, but you could have more than that. For example the `min-height` fix isn't doing to work if someone has 23 Topics. We will want to add an increased bottom margin or padding, whichever you're using to separate the topic items. I would suggest that the distance of the last topic and the footer be equal to the distance between the avatar and the first topic in the list.

File: root/static/js/userpages/UserPage.js [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        var randomNum = Math.ceil(Math.random() * 1e7);
+        pre_id = pre_id? '_' + pre_id : '';
+        element_id = element_id? '_' + element_id.replace(/[^0-9A-Za-z]/g, '') : '';
+        var imgUrlSuffix = pre_id + element_id + '?' + randomNum;
+
+        $scope.newImg = new Image();
+        $scope.newImg.src = $scope.imgUrl + imgUrlSuffix;
+	};
+
+	$scope.ias = [];
```
Comment (alohaas) [inline root/static/js/userpages/UserPage.js]: Indentation starts to get a little funky here through line 111.

File: root/static/js/userpages/UserPage.js [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	// given an array, search for an attr (like 'username'): if found, return object
+	findByAttr: function(arr, attr, name) {
+	    return _.find(arr, function(obj) {
+		return obj[attr] == name;
+	    });
+	},
+
+    // return the first char in a dev's username (for when avatar is not available)
+    firstUsernameChar: function(username) {
+        username = username? username.charAt(0) : '';
```
Comment (alohaas) [inline root/static/js/userpages/UserPage.js]: nitpick but that `?` needs a space before it :grimacing:

**Comment by jagtalon** [comment]
@alohaas auto-indented and fixed the space

**Comment by GuiltyDolphin** [comment]
@jagtalon @alohaas When I refresh the page on https://duck.co/u/guiltydolphin/ I very briefly see a menu appear to the right, this is the best screenshot I've managed to get of it:

screenshot from 2016-04-13 18 12 54

I don't see any way to expand it - any ideas?

**Comment by alohaas** [comment]
@GuiltyDolphin thanks for pointing that out – it's a vestige from an earlier version of the design. We'll be removing it.

**Comment by jagtalon** [comment]
@GuiltyDolphin thanks! What @alohaas said. I'll shake it off from the code

giphy

**Comment by jagtalon** [comment]
@alohaas removed the slideout, although still figuring out the min-height thing. Can I postpone that fix?

**Comment by alohaas** [comment]
@jagtalon I think it's OK if we postpone the fix. Next round of larger changes might warrant a revisit.

**Comment by nilnilnil** [comment]
What is the testing strategy for this? How will we validate correctness in the future?

**Pull request merged**